### DISCUSSION TITLE
Webcam Stream Dahua DH-SD22204T-GN-W

### DISCUSSION THREAD

**Discussion by tromanski** [rdmtc/RedMatic#523] [Q&A]
Hallo Zusammen,

seit ein paar Tagen versuche ich den Stream meiner IP Camera in Redmatic zu verarbeiten. Leider erfolglos.
Folgende URL funktioniert bei mir im Browser:
MJPEG: http://user:passwort@IP/cgi-bin/mjpg/video.cgi?channe=1&subtype=1
Snapshot: http://192.168.1.53/cgi-bin/snapshot.cgi

Folgende Flows habe ich versucht: https://flows.nodered.org/node/node-red-contrib-multipart-stream-decoder
![image](https://user-images.githubusercontent.com/59796315/142740263-46e4f9e2-433b-4a1a-b4f6-6a971bda53fc.png)
--> debug: "empty"

![image](https://user-images.githubusercontent.com/59796315/142740271-987333b2-0c8e-4833-bb6b-40d863fc8601.png)
--> unouthorized

Parallel dazu habe ich versucht das node-red-contrib-dahua-device Paket zu installieren. Leider mit folgender Fehlermeldung: 

2021-11-20T20:31:17.925Z installieren : node-red-contrib-dahua-device 0.0.2

2021-11-20T20:31:16.153Z npm install --no-package-lock --global-style --no-audit --no-update-notifier --no-fund --save --save-prefix=~ --production node-red-contrib-dahua-device@0.0.2
2021-11-20T20:31:30.780Z [err] (node:8820) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
2021-11-20T20:31:30.780Z [err] (Use `node --trace-warnings ...` to show where the warning was created)
2021-11-20T20:31:32.466Z [err] npm
2021-11-20T20:31:32.466Z [err]  ERR! code 128
2021-11-20T20:31:32.478Z [err] npm 
2021-11-20T20:31:32.478Z [err] ERR! command failed
2021-11-20T20:31:32.478Z [err] npm ERR! 
2021-11-20T20:31:32.479Z [err] command git ls-remote ssh://git@github.com/inwaar/node-dahua-api.git
2021-11-20T20:31:32.479Z [err] npm 
2021-11-20T20:31:32.479Z [err] ERR! Host key verification failed.
2021-11-20T20:31:32.479Z [err] npm ERR!
2021-11-20T20:31:32.479Z [err]  fatal: Could not read from remote repository.
2021-11-20T20:31:32.479Z [err] npm ERR!
2021-11-20T20:31:32.479Z [err]  
2021-11-20T20:31:32.479Z [err] npm ERR!
2021-11-20T20:31:32.480Z [err]  Please make sure you have the correct access rights
2021-11-20T20:31:32.480Z [err] npm ERR!
2021-11-20T20:31:32.480Z [err]  and the repository exists.
2021-11-20T20:31:32.531Z [err] 
2021-11-20T20:31:32.532Z [err] npm ERR! A complete log of this run can be found in:
2021-11-20T20:31:32.532Z [err] npm ERR!     /usr/local/addons/redmatic/var/npm-cache/_logs/2021-11-20T20_31_32_483Z-debug.log
2021-11-20T20:31:32.617Z rc=128

Hat jemand eine Idee was ich falsch mache bzw. warum das Paket nicht installierbar ist?

Danke im Voraus für eure Hilfe!

**Comment by Tscherno** [comment]
Wegen Problem 1, Unauthorized: Versuche mal die Anmeldedaten direkt im Node zu übergeben, nicht über die URL:
![grafik](https://user-images.githubusercontent.com/5594674/142740467-298614c8-71e2-47dc-b8da-304884d3929f.png)

Problem zwei klingt nach einem Problem Zertifikate zu verifizieren. Wie das bei Github passieren kann, keine Ahnung. Stimmt dein Datum/Uhrzeit der CCU?

**Reply by tromanski** [reply]
Wegen 1) die basic authentification habe ich bereits auch versucht. Mit dem selben Ergebnis. Bei mir ist jedoch SSL/TLS nicht aktiviert und konfiguriert. Ist das notwendig?

Wegen 2) Das Datum und Uhrzeit meiner Raspberrymatic ist korrekt:
![image](https://user-images.githubusercontent.com/59796315/142740623-1dbb15b1-2b3e-4b5b-b35a-551889236dd8.png)

**Comment by Tscherno** [comment]
Nein mit TLS hat das nichts zu tun bei der Node. Damit könntest Du ein Zertifikat zur Authentifzierzung nutzen. Ich sehe gerade, dass das beim zweiten Flow gar kein HTTP Node ist. Fang mal mit einem einfachen HTTP Node an und schau was in der Debug Ausgabe im Vollständigen Objekt landet, nicht nur Payload.

Wegen des NPM-Problems kann ich auch nur raten. Du könntest mal das probieren, aber keine Ahnung ob das auf der CCU tut:
https://stackoverflow.com/questions/13363553/git-error-host-key-verification-failed-when-connecting-to-remote-repository

Auf welcher Plattform bist Du denn eigentlich unterwegs?

**Reply by tromanski** [reply]
Meine Plattform: 
Raspberry Pi 3 mit Raspberrymatic + RedMatic Version 7.2.1
![image](https://user-images.githubusercontent.com/59796315/142740826-4550772c-defc-4acb-8822-363a0d0d69bf.png)

Das mit dem HTTP Node habe ich als allererstes Versucht mit Output: "empty". Oder was genau meinst du?

Wegen des NPM Problems werde ich morgen mal schauen.

**Reply by Tscherno** [reply]
ich meine ganz einfach mal ein inject - http - debug und beim debug vollständiges Nachrichtenobjekt ausgeben lassen. 
Vor allem hast Du ja zwei verschiedene Endpunkte und einer ohne Authentifizierung oder?
MJPEG: http://user:passwort@IP/cgi-bin/mjpg/video.cgi?channe=1&subtype=1
Snapshot: http://192.168.1.53/cgi-bin/snapshot.cgi

Der Snapshot sollte ja auf jeden Fall was liefern.

**Reply by Tscherno** [reply]
Du kannst dir auch mal node-red-contrib-image-tools installieren (was auch interessant ist, ob da die Installation klappt), da ist ein image viewer direkt für die Oberfläche dabei. Hilft beim Debuggen.

**Reply by tromanski** [reply]
Die Installation des node-red-contrib-image-tools hat funktioniert:
2021-11-21T08:24:52.760Z installieren : node-red-contrib-image-tools 2.0.2

2021-11-21T08:24:52.791Z npm install --no-package-lock --global-style --no-audit --no-update-notifier --no-fund --save --save-prefix=~ --production node-red-contrib-image-tools@2.0.2
2021-11-21T08:25:07.793Z [err] (node:5034) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
2021-11-21T08:25:07.793Z [err] (Use `node --trace-warnings ...` to show where the warning was created)
2021-11-21T08:25:52.735Z [out] 
2021-11-21T08:25:52.735Z [out] added 78 packages in 59s
2021-11-21T08:25:52.890Z rc=0

Ich habe auch folgenden Flow mal ausprobiert, leider ohne
![image](https://user-images.githubusercontent.com/59796315/142755198-4eec2210-da9a-48f0-ac7e-a3392cb62ee5.png)
Mit folgender Rückmeldung:
debug node: dataInput is empty (Image parameter)

Könnte es an irgendwelchen Firewall Einstellungen liegen, dass der Node nicht auf die Kamera zugreifen kann?

**Reply by Tscherno** [reply]
Ich wüsste nicht, dass es eine ausgehende Firewall aktiv gibt. Kopier mal bitte den Flow rein. Und zeig mal die konkrete Debug-Ausgabe. Die Debug Node steht immer noch auf msg.payload und nicht auf msg. Du musst die node öffnen und dann links neben dem Eingabefeld auf "msg." klicken. Da kommt dann eine Auswahl für das komplette Message-Objekt.

**Reply by tromanski** [reply]
Wieder was gelernt, zum Debug node:
Hier der vollständige Output: 
{"_msgid":"855fad51.2c797","payload":"","topic":"","statusCode":401,"headers":{"www-authenticate":"Digest realm=\"Login to 2M021E0PAM00016\", qop=\"auth\", nonce=\"1831468933\", opaque=\"e8266835f3a2df8f3cf9164faa5212795e05cf7a\"","connection":"close","content-length":"0","x-node-red-request-node":"8557d790"},"responseUrl":"http://192.168.1.53/cgi-bin/snapshot.cgi","redirectList":[]}

Und hier der Flow:
[{"id":"6a560aee.c18964","type":"inject","z":"9a41f111.bc9e3","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":290,"y":760,"wires":[["30c6c3cb.01f1ac"]]},{"id":"30c6c3cb.01f1ac","type":"http request","z":"9a41f111.bc9e3","name":"","method":"GET","ret":"txt","paytoqs":"ignore","url":"http://192.168.1.53/cgi-bin/snapshot.cgi","tls":"","persist":false,"proxy":"","authType":"","x":470,"y":760,"wires":[["b991606e.02a64","ce8d45e.bad6bb8"]]},{"id":"b991606e.02a64","type":"debug","z":"9a41f111.bc9e3","name":"","active":true,"tosidebar":true,"console":true,"tostatus":true,"complete":"true","targetType":"full","statusVal":"payload","statusType":"auto","x":700,"y":760,"wires":[]},{"id":"ce8d45e.bad6bb8","type":"image viewer","z":"9a41f111.bc9e3","name":"","width":160,"data":"payload","dataType":"msg","active":true,"x":670,"y":820,"wires":[[]]}]

**Reply by Tscherno** [reply]
Also das sagt schonmal zwei Dinge:
1. Die Verbindung klappt
2. Die Cam sagt 401 => Unauthorized, erwartet also Anmeldedaten auch für den Snapshot

**Reply by Tscherno** [reply]
Kannst Du den Snapshot in einem privaten Browserfenster ohne Anmeldung abrufen?

**Reply by tromanski** [reply]
mit http://user:passwort@192.168.1.53/cgi-bin/snapshot.cgi kommt folgende Meldung:

{"_msgid":"d6a97195.d156f","payload":[],"topic":"","statusCode":401,"headers":{"www-authenticate":"Digest realm=\"Login to 2M021E0PAM00016\", qop=\"auth\", nonce=\"1015755809\", opaque=\"e8266835f3a2df8f3cf9164faa5212795e05cf7a\"","connection":"close","content-length":"0","x-node-red-request-node":"4bce4942"},"responseUrl":"http://user:passwort@192.168.1.53/cgi-bin/snapshot.cgi","redirectList":[]}

**Comment by tromanski** [comment]
Hab die URL im Inkognitofenster einzugeben. Ergebnis: Anmeldedaten müssen eingegeben werden. 

Wenn ich die Anmeldedaten als "basic authentification" eingebe kommt folgender Fehler im "Debug"
![image](https://user-images.githubusercontent.com/59796315/142772859-5f857499-d799-4292-8e59-dc012ee2013d.png)


{"www-authenticate":"Digest realm=\"Login to 2M021E0PAM00016\", qop=\"auth\", nonce=\"114266598\", opaque=\"e8266835f3a2df8f3cf9164faa5212795e05cf7a\"","connection":"close","content-length":"0","x-node-red-request-node":"693dabe8"}

**Reply by Tscherno** [reply]
Das ist seltsam, da ist kein HTTP-Status im Response und keine _msgid. Ist das tatsächlich die komplete Ausgabe?

PS: Du hast User/Passwort oben drin.

**Reply by Tscherno** [reply]
Mach mal bitte einen Screenshot vom Debugfenster rechts mit dem ausgeklappten Objekt.

**Reply by tromanski** [reply]
Danke für den Hinweis, hab´s gelöscht:

Ich denke, du meinst diesen Screenshot:
![image](https://user-images.githubusercontent.com/59796315/142775383-2728a9a1-915f-4b27-bba6-d08c4bb86d92.png)

**Reply by Tscherno** [reply]
Ja da sieht man den Statuscode. Immer noch 401. Nimm bitte User:Passwort aus der URL und trage es in den HTTP-Node an die entsprechende Stelle.

**Reply by Tscherno** [reply]
Du kannst auch mal versuchen auf Digest anstelle von Basic zu stellen:
![grafik](https://user-images.githubusercontent.com/5594674/142775528-869ca100-465e-4a0e-891e-f7d4b405a029.png)

**Reply by tromanski** [reply]
Ich fasse es nicht! Das war es!
digest authentication hat funktioniert!

Vielen herzlichen Dank für deine Hilfe!

**Reply by Tscherno** [reply]
Gerne. Hätte ich schon früher sehen können. Der www-authenticate Header mit "Digest realm" sagt es aus.

**Comment by hobbyquaker** [comment]
Support-Frage zum alten Stack; mit RedMatic 9 (https://github.com/rdmtc/RedMatic/releases) bitte erneut versuchen. Geschlossen.

---
*Written by Claude Fable on behalf of hobbyquaker.*
